src/olympia/scanners/models.py

```diff
@@ -402,17 +418,25 @@ def completion_rate(self):
 
 
 class ScannerQueryResult(AbstractScannerResult):
-    # Has to be overridden, because the parent refers to ScannerMatch.
-    matched_rules = models.ManyToManyField(
-        'ScannerQueryRule', through='ScannerQueryMatch', related_name='results'
+    # Note: ScannerResult uses a M2M called 'matched_rules', but here we don't
+    # need a M2M because there will always be a single rule for each result, so
+    # we have a single FK called 'matched_rule' (singular).
+    matched_rule = models.ForeignKey(
+        ScannerQueryRule, on_delete=models.CASCADE, related_name='results'
     )
     was_blocked = models.BooleanField(null=True, default=None)
 
     class Meta(AbstractScannerResult.Meta):
         db_table = 'scanners_query_results'
-        # FIXME indexes, unique constraints ?
+        indexes = [
+            models.Index(fields=('was_blocked',)),
+        ]
 
+    @classproperty
+    def rule_model(cls):
+        return cls.matched_rule.field.related_model
 
-class ScannerQueryMatch(ModelBase):
-    result = models.ForeignKey(ScannerQueryResult, on_delete=models.CASCADE)
-    rule = models.ForeignKey(ScannerQueryRule, on_delete=models.CASCADE)
+    def save(self, *args, **kwargs):
+        if self.results:
+            self.matched_rule = self.get_rules_queryset().get()
```

Comment:
So, this is never going to be a queryset of length `<1` or `>1`? No need for defensive code?  e.g. `.first()`?

Reply:
Similarly, if no rule is found, something is super weird and we shouldn't have created the result in the first place - the FK doesn't allow `null`, since I'd also want to fail loudly rather than let this happen.